recirq/quantum_chess/quantum_board.py

```diff
@@ -331,19 +369,64 @@ def _generate_accumulations(self, repetitions: int = 1000) -> None:
 
         self.accumulations_repetitions = repetitions
 
+    def _caching_supported(self, m: move.Move):
+        """Checks if caching is supported for this move."""
+
+        # Caching is supported for a split jump from one full square to two empty squares.
+        if m.move_type == enums.MoveType.SPLIT_JUMP and nth_bit_of(square_to_bit(m.source), self.full_squares)\
```

Comment:
self.full/empty squares are only available if a sample has been taken, which is not always the state. Instead use `self.entangled_squares` (which squares are not known to be in a classical state), and `self.state` which tracks occupancy for non-entangled squares. See [here](https://github.com/quantumlib/ReCirq/blob/4431f47bc0bf6f87a95adfbe0674026f8cd1506a/recirq/quantum_chess/quantum_board.py#L586) for an example.

Reply:
Checking entangled_squares doesn't seem to work because do_move is usually called before getting the probability distribution, and do_move already adds the targets in the split_jump to entangled squares. Do you think I can just set an initial value for full and empty squares?

Reply:
no I mean, the do_move already modifies `entangled_squares`, so we can't use it to check whether the target is empty or entangled. or am I misunderstanding?

Reply:
Oh yes, you are right. I suppose you could work around this by either creating arrays of full/empty squares as known from `entangled_squares`/`state` updated after each move. Or alternatively, determine whether caching is supported inside of `do_move` and store that boolean somewhere.

Reply:
For all uses of the cache apart from the first move, we will need to have samples for the previous move anyway, since we need the prob distribution from the previous move to use the cache, so it won't be a problem. For the first move, I can just set the full_squares and empty_squares based on the init_basis_state (like what I did in this PR). It seems to work, do you see any problems?

Reply:
OK, sounds good to me. Although the answer to whether a move should be cached could be "wrong" in principle, but the results would be off only a small amount and it would not change the circuit.